internal/controllers/extension_controller.go

```diff
@@ -123,22 +147,69 @@ func (r *ExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alpha1.Ext
+	bundle, err := r.resolve(ctx, *ext)
+	if err != nil {
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as resolution failed", ext.GetGeneration())
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as resolution failed", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
+
+	// Now we can set the Resolved Condition, and the resolvedBundleSource field to the bundle.Image value.
+	ext.Status.ResolvedBundleResource = bundle.Image
+	setResolvedStatusConditionSuccess(&ext.Status.Conditions, fmt.Sprintf("resolved to %q", bundle.Image), ext.GetGeneration())
 
-	// Set the TypeInstalled condition to Unknown to indicate that the resolution
-	// hasn't been attempted yet, due to the spec being invalid.
-	ext.Status.InstalledBundleResource = ""
-	setInstalledStatusConditionUnknown(&ext.Status.Conditions, "the Extension interface is not fully implemented", ext.GetGeneration())
-	// Set the TypeResolved condition to Unknown to indicate that the resolution
-	// hasn't been attempted yet, due to the spec being invalid.
-	ext.Status.ResolvedBundleResource = ""
-	setResolvedStatusConditionUnknown(&ext.Status.Conditions, "the Extension interface is not fully implemented", ext.GetGeneration())
+	mediaType, err := bundle.MediaType()
+	if err != nil {
+		setInstalledStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as installation has failed", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
 
-	setDeprecationStatusesUnknown(&ext.Status.Conditions, "the Extension interface is not fully implemented", ext.GetGeneration())
+	// TODO: this needs to include the registryV1 bundle option. As of this PR, this only supports direct
+	// installation of a set of manifests.
+	if mediaType != catalogmetadata.MediaTypePlain {
+		// Set the TypeInstalled condition to Unknown to indicate that the resolution
+		// hasn't been attempted yet, due to the spec being invalid.
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, fmt.Sprintf("bundle type %s not supported currently", mediaType), ext.GetGeneration())
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as installation has failed", ext.GetGeneration())
+		return ctrl.Result{}, nil
+	}
+
+	app := r.GenerateExpectedApp(*ext, bundle.Image)
+	if err := r.ensureApp(ctx, app); err != nil {
+		// originally Reason: ocv1alpha1.ReasonInstallationFailed
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
```

Comment:
What if a previous `App` _is_ installed? It seems like we should be able to keep this status reflecting the truth of what's actually installed? This can be a follow-up.

Reply:
If existing app with the same spec exists, we will be returning with a nil error after doing the equality check: https://github.com/operator-framework/operator-controller/pull/625/files#diff-c04822d94b6cb8ff190859213d75fa6cd5c8b7c5284aaa2cf93cd504f727f5a9R363. After which we map the status of App we return from `ensureApp` - which is going to be either the existing one or the patched one.

Reply:
But if we have a new spec (e.g. we're doing an upgrade), and we patch the `App`, does the `App` status have a dual indication of:
- Still have this old thing active
- Progressing on rolling out the new thing

It would be nice, for example, that if something failed during one of (fetch, template, preflight) we could tell Extension users that the previous version is still installed, but this upgrade needs some attention.

Reply:
> we could tell Extension users that the previous version is still installed, but this upgrade needs some attention.

Upgrade scenario needs to be verified with kapp-controller. I think the App status still reflects "Reconciling", I don't see anything specific related to it in their status. Will create a follow up to look into it.